File: JSPM.Web/Modules/Syllabus/Chapters/ChaptersRow.cs
using Serenity.ComponentModel;
using Serenity.Data;
using Serenity.Data.Mapping;
using System.ComponentModel;

namespace JSPM.Syllabus;

[ConnectionKey("Default"), Module("Syllabus"), TableName("[Syllabus].[Chapters]")]
[DisplayName("Chapters"), InstanceName("Chapters")]
[ReadPermission("Administration:General")]
[ModifyPermission("Administration:General")]
public sealed class ChaptersRow : Row<ChaptersRow.RowFields>, IIdRow, INameRow
{
    const string jSubject = nameof(jSubject);

    [DisplayName("Id"), Identity, IdProperty]
    public int? Id
    {
        get => fields.Id[this];
        set => fields.Id[this] = value;
    }

    [DisplayName("Subject"), ForeignKey("[Syllabus].[Subjects]", "Id"), LeftJoin(jSubject), TextualField(nameof(SubjectName))]
    public int? SubjectId
    {
        get => fields.SubjectId[this];
        set => fields.SubjectId[this] = value;
    }

    [DisplayName("Chapter Name"), Size(200), NotNull, QuickSearch, NameProperty]
    public string ChapterName
    {
        get => fields.ChapterName[this];
        set => fields.ChapterName[this] = value;
    }

    [DisplayName("Description"), Size(1000)]
    public string Description
    {
        get => fields.Description[this];
        set => fields.Description[this] = value;
    }

    [DisplayName("Index")]
    public int? Index
    {
        get => fields.Index[this];
        set => fields.Index[this] = value;
    }

    [DisplayName("Subject Subject Name"), Expression($"{jSubject}.[Subject Name]")]
    public string SubjectName
    {
        get => fields.SubjectName[this];
        set => fields.SubjectName[this] = value;
    }

    public class RowFields : RowFieldsBase
    {
        public Int32Field Id;
        public Int32Field SubjectId;
        public StringField ChapterName;
        public StringField Description;
        public Int32Field Index;

        public StringField SubjectName;
    }
}

File: JSPM.Web/Modules/ServerTypes/Syllabus/ChaptersRow.ts
import { fieldsProxy } from "@serenity-is/corelib/q";

export interface ChaptersRow {
    Id?: number;
    SubjectId?: number;
    ChapterName?: string;
    Description?: string;
    Index?: number;
    SubjectName?: string;
}

export abstract class ChaptersRow {
    static readonly idProperty = 'Id';
    static readonly nameProperty = 'ChapterName';
    static readonly localTextPrefix = 'Syllabus.Chapters';
    static readonly deletePermission = 'Administration:General';
    static readonly insertPermission = 'Administration:General';
    static readonly readPermission = 'Administration:General';
    static readonly updatePermission = 'Administration:General';

    static readonly Fields = fieldsProxy<ChaptersRow>();
}